File: sample_apps_reinvent2021/python/examples/ScheduledQueryExample.py
#!/usr/bin/python

import json
import time
from string import ascii_lowercase
from random import choice

from utils.WriteUtil import WriteUtil
from utils.QueryUtil import QueryUtil
from utils.TimestreamDependencyHelper import TimestreamDependencyHelper


class ScheduledQueryExample:
    HOSTNAME = "host-24Gju"
    SQ_NAME = "daily-sample"
    TOPIC = "sq_sample_app_topic"
    QUEUE_NAME = "sq_sample_app_queue"
    ROLE_NAME = "ScheduledQuerySampleApplicationRole"
    POLICY_NAME = "ScheduledQuerySampleApplicationAccess"
    ERROR_BUCKET_NAME = "scheduledquerysamplerrorbucket" + ''.join([choice(ascii_lowercase) for _ in range(5)])
    SQ_ENABLED_STATE = "ENABLED"
    SQ_DISABLED_STATE = "DISABLED"

    policy_arn = ""
    subscription_arn = ""
    queue_url = ""
    scheduled_query_arn = ""
    s3_error_bucket = ""

    def __init__(
            self,
            stage,
            region,
            database_name,
            table_name,
            sq_database_name,
            sq_table_name,
            write_client,
            query_client,
            skip_deletion,
            fail_on_execution=False
    ):
        self.write_client = write_client
        self.stage = stage
        self.region = region
        self.query_client = query_client
        self.table_name = table_name
        self.database_name = database_name
        self.sq_database_name = sq_database_name
        self.sq_table_name = sq_table_name
        self.skip_deletion = skip_deletion

        # Find the average, p90, p95, and p99 CPU utilization for a specific EC2 host over the past 2 hours.
        self.valid_query = \
            "SELECT region, az, hostname, BIN(time, 15s) AS binned_timestamp, " \
            "    ROUND(AVG(cpu_utilization), 2) AS avg_cpu_utilization, " \
            "    ROUND(APPROX_PERCENTILE(cpu_utilization, 0.9), 2) AS p90_cpu_utilization, " \
            "    ROUND(APPROX_PERCENTILE(cpu_utilization, 0.95), 2) AS p95_cpu_utilization, " \
            "    ROUND(APPROX_PERCENTILE(cpu_utilization, 0.99), 2) AS p99_cpu_utilization " \
            "FROM " + database_name + "." + table_name + " " \
                                                         "WHERE measure_name = 'metrics' " \
                                                         "AND hostname = '" + self.HOSTNAME + "' " \
                                                                                              "AND time > ago(2h) " \
                                                                                              "GROUP BY region, hostname, az, BIN(time, 15s) " \
                                                                                              "ORDER BY binned_timestamp ASC " \
                                                                                              "LIMIT 5"
        self.invalid_query = "SELECT cast('2030-12-16' as TIMESTAMP) as timestamp, 1.0 as random_measure_value, '1' as dim0"
        self.fail_on_execution = fail_on_execution

    def create_scheduled_query_helper(self, topic_arn, role_arn, query, target_configuration):
        print("\nCreating Scheduled Query")
        schedule_configuration = {
            'ScheduleExpression': 'cron(0/2 * * * ? *)'
        }
        notification_configuration = {
            'SnsConfiguration': {
                'TopicArn': topic_arn
            }
        }
        error_report_configuration = {
            'S3Configuration': {
                'BucketName': self.s3_error_bucket
            }
        }

        try:
            create_scheduled_query_response = \
                self.query_client.create_scheduled_query(Name=self.SQ_NAME,
                                                         QueryString=query,
                                                         ScheduleConfiguration=schedule_configuration,
                                                         NotificationConfiguration=notification_configuration,
                                                         TargetConfiguration=target_configuration,
                                                         ScheduledQueryExecutionRoleArn=role_arn,
                                                         ErrorReportConfiguration=error_report_configuration
                                                         )
            print("Successfully created scheduled query : ", create_scheduled_query_response['Arn'])
            return create_scheduled_query_response['Arn']
        except Exception as err:
            print("Scheduled Query creation failed:", err)
            raise err

    def create_valid_scheduled_query(self, topic_arn, role_arn):
        target_configuration = {
            'TimestreamConfiguration': {
                'DatabaseName': self.sq_database_name,
                'TableName': self.sq_table_name,
                'TimeColumn': 'binned_timestamp',
                'DimensionMappings': [
                    {'Name': 'region', 'DimensionValueType': 'VARCHAR'},
                    {'Name': 'az', 'DimensionValueType': 'VARCHAR'},
                    {'Name': 'hostname', 'DimensionValueType': 'VARCHAR'}
                ],
                'MultiMeasureMappings': {
                    'TargetMultiMeasureName': 'target_name',
                    'MultiMeasureAttributeMappings': [
                        {'SourceColumn': 'avg_cpu_utilization', 'MeasureValueType': 'DOUBLE',
                         'TargetMultiMeasureAttributeName': 'avg_cpu_utilization'},
                        {'SourceColumn': 'p90_cpu_utilization', 'MeasureValueType': 'DOUBLE',
                         'TargetMultiMeasureAttributeName': 'p90_cpu_utilization'},
                        {'SourceColumn': 'p95_cpu_utilization', 'MeasureValueType': 'DOUBLE',
                         'TargetMultiMeasureAttributeName': 'p95_cpu_utilization'},
                        {'SourceColumn': 'p99_cpu_utilization', 'MeasureValueType': 'DOUBLE',
                         'TargetMultiMeasureAttributeName': 'p99_cpu_utilization'},
                    ]
                }
            }
        }

        return self.create_scheduled_query_helper(topic_arn, role_arn, self.valid_query, target_configuration)

    def create_invalid_scheduled_query(self, topic_arn, role_arn):
        target_configuration = {
            'TimestreamConfiguration': {
                'DatabaseName': self.sq_database_name,
                'TableName': self.sq_table_name,
                'TimeColumn': 'timestamp',
                'DimensionMappings': [
                    {'Name': 'dim0', 'DimensionValueType': 'VARCHAR'}
                ],
                'MultiMeasureMappings': {
                    'TargetMultiMeasureName': 'target_name',
                    'MultiMeasureAttributeMappings': [
                        {'SourceColumn': 'random_measure_value', 'MeasureValueType': 'DOUBLE'},
                    ]
                }
            }
        }

        return self.create_scheduled_query_helper(topic_arn, role_arn, self.invalid_query, target_configuration)

    def list_scheduled_queries(self):
        print("\nListing Scheduled Queries")
        try:
            response = self.query_client.list_scheduled_queries(MaxResults=10)
            self.print_scheduled_queries(response['ScheduledQueries'])
            next_token = response.get('NextToken', None)
            while next_token:
                response = self.query_client.list_scheduled_queries(NextToken=next_token, MaxResults=10)
                self.print_scheduled_queries(response['ScheduledQueries'])
                next_token = response.get('NextToken', None)
        except Exception as err:
            print("List scheduled queries failed:", err)
            raise err

    def execute_scheduled_query(self, scheduled_query_arn, invocation_time):
        print("\nExecuting Scheduled Query")
        try:
            self.query_client.execute_scheduled_query(ScheduledQueryArn=scheduled_query_arn, InvocationTime=invocation_time)
            print("Successfully started executing scheduled query")
        except self.query_client.exceptions.ResourceNotFoundException as err:
            print("Scheduled Query doesn't exist")
            raise err
        except Exception as err:
            print("Scheduled Query execution failed:", err)
            raise err

    def delete_scheduled_query(self, scheduled_query_arn):
        print("\nDeleting Scheduled Query")
        try:
            self.query_client.delete_scheduled_query(ScheduledQueryArn=scheduled_query_arn)
            print("Successfully deleted scheduled query :", scheduled_query_arn)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Scheduled Query deletion failed:", err)

    def describe_scheduled_query(self, scheduled_query_arn):
        print("\nDescribing Scheduled Query")
        try:
            response = self.query_client.describe_scheduled_query(ScheduledQueryArn=scheduled_query_arn)
            if 'ScheduledQuery' in response:
                response = response['ScheduledQuery']
                for key in response:
                    print("{} :{}".format(key, response[key]))
        except self.query_client.exceptions.ResourceNotFoundException as err:
            print("Scheduled Query doesn't exist")
            raise err
        except Exception as err:
            print("Scheduled Query describe failed:", err)
            raise err

    def update_scheduled_query(self, scheduled_query_arn, state):
        print("\nUpdating Scheduled Query")
        try:
            self.query_client.update_scheduled_query(ScheduledQueryArn=scheduled_query_arn,
                                                     State=state)
            print("Successfully update scheduled query state to", state)
        except self.query_client.exceptions.ResourceNotFoundException as err:
            print("Scheduled Query doesn't exist")
            raise err
        except Exception as err:
            print("Scheduled Query deletion failed:", err)
            raise err

    def write_records(self):
        print("Writing records")
        current_time = WriteUtil.current_milli_time()

        dimensions = [
            {'Name': 'region', 'Value': 'us-east-1'},
            {'Name': 'az', 'Value': 'az1'},
            {'Name': 'hostname', 'Value': 'host-24Gju'}
        ]

        common_attributes = {
            'Dimensions': dimensions,
            'Time': current_time
        }

        cpu_utilization = {
            'Name': 'cpu_utilization',
            'Value': '13.6',
            'Type': 'DOUBLE',
        }

        memory_utilization = {
            'Name': 'memory_utilization',
            'Value': '40',
            'Type': 'DOUBLE',
        }

        records = [{
            'MeasureName': 'metrics',
            'MeasureValues': [cpu_utilization, memory_utilization],
            'MeasureValueType': 'MULTI'
        }]

        try:
            result = self.write_client.write_records(DatabaseName=self.database_name, TableName=self.table_name,
                                                     Records=records, CommonAttributes=common_attributes)
            if result and result['ResponseMetadata']:
                print("WriteRecords Status: [%s]" % result['ResponseMetadata']['HTTPStatusCode'])
        except self.write_client.exceptions.RejectedRecordsException as err:
            WriteUtil.print_rejected_records_exceptions(err)
        except Exception as err:
            print("Error:", err)

    @staticmethod
    def print_scheduled_queries(scheduled_queries):
        for scheduled_query in scheduled_queries:
            print(scheduled_query['Arn'])

    def parse_s3_error_report(self, bucket, prefix):
        objects = TimestreamDependencyHelper().list_s3_objects(bucket, prefix)

        for o in objects:
            body = o.get()['Body'].read().decode('utf-8')
            print(f"error: {body}")

    def run(self):
        write_util = WriteUtil(self.write_client)
        query_util = QueryUtil(self.query_client, self.database_name, self.table_name)
        timestream_dependency_helper = TimestreamDependencyHelper(self.region)

        try:
            # Create base table and ingest records
            write_util.create_database(self.database_name)
            write_util.create_table(self.database_name, self.table_name)
            self.write_records()

            # Create database and table to store scheduled query results
            write_util.create_database(self.sq_database_name)
            write_util.create_table(self.sq_database_name, self.sq_table_name)

            # Create sns, sqs for scheduled query
            self.topic_arn = timestream_dependency_helper.create_sns(self.TOPIC)
            self.queue_url = timestream_dependency_helper.create_sqs_queue(self.QUEUE_NAME)
            # Need to wait at least 1s after creating queue to use it
            time.sleep(2)
            self.queue_arn = timestream_dependency_helper.get_queue_arn(self.queue_url)
            self.subscription_arn = timestream_dependency_helper.subscribe_to_sns_topic(self.topic_arn, self.queue_arn)
            timestream_dependency_helper.set_sqs_access_policy(self.queue_url, self.queue_arn, self.topic_arn)

            # Create role for scheduled query access
            self.role_arn = timestream_dependency_helper.create_role(self.ROLE_NAME, self.stage, self.region)
            self.policy_arn = timestream_dependency_helper.create_policy(self.POLICY_NAME)
            timestream_dependency_helper.attach_policy_to_role(self.ROLE_NAME, self.policy_arn)

            # Waiting for newly created role to be active
            print("\nWaiting for newly created role to become active")
            time.sleep(15)

            # Create S3 bucket to hold error reports
            self.s3_error_bucket = timestream_dependency_helper.create_s3_bucket(self.ERROR_BUCKET_NAME)

            # Scheduled Query Activities
            if self.fail_on_execution:
                self.scheduled_query_arn = self.create_invalid_scheduled_query(self.topic_arn, self.role_arn)
            else:
                self.scheduled_query_arn = self.create_valid_scheduled_query(self.topic_arn, self.role_arn)

            self.describe_scheduled_query(self.scheduled_query_arn)
            self.execute_scheduled_query(self.scheduled_query_arn, time.time())
            time.sleep(15)
            print("Waiting for scheduled query to complete execution")

            # polling sqs queue for notification
            message = {}
            response = ''
            for _ in range(10):
                message = timestream_dependency_helper.receive_message(self.queue_url)
                response = message['MessageAttributes']['notificationType']['Value']
                if not (response == 'SCHEDULED_QUERY_CREATING' or response == 'SCHEDULED_QUERY_CREATED'):
                    break

            if response == 'MANUAL_TRIGGER_SUCCESS':
                print("Received notification regarding scheduled query execution")
                print("Fetching Scheduled Query execution results")
                query_util.run_query("SELECT * FROM " + self.sq_database_name + "." + self.sq_table_name)
            elif response == "MANUAL_TRIGGER_FAILURE":
                # print error report in bucket
                message_json = json.loads(message['Message'])
                schedule_query_run_summary = message_json['scheduledQueryRunSummary']
                print(f"Failure Reason from SQS Notification:: {schedule_query_run_summary['failureReason']}")

                if 'errorReportLocation' in schedule_query_run_summary:
                    error_report_location = schedule_query_run_summary['errorReportLocation']
                    object_key = error_report_location['s3ReportLocation']['objectKey']
                    self.parse_s3_error_report(self.s3_error_bucket, object_key)
            else:
                print("UNEXPECTED ERROR: Expected notification MANUAL_TRIGGER_SUCCESS, Received ", response)

            self.list_scheduled_queries()
            self.update_scheduled_query(self.scheduled_query_arn, self.SQ_DISABLED_STATE)
        finally:
            if not self.skip_deletion:
                self.cleanup(write_util, timestream_dependency_helper)

    def cleanup(self, write_util, scheduled_query_util):
        print("cleaning up resources created by sample")
        if self.scheduled_query_arn:
            self.delete_scheduled_query(self.scheduled_query_arn)

        # Clean up for scheduled query
        scheduled_query_util.detach_policy_from_role(self.ROLE_NAME, self.policy_arn)
        scheduled_query_util.delete_policy(self.policy_arn)
        scheduled_query_util.delete_role(self.ROLE_NAME)
        scheduled_query_util.unsubcribe_from_sns(self.subscription_arn)
        scheduled_query_util.delete_sqs_queue(self.queue_url)
        scheduled_query_util.delete_sns(self.topic_arn)
        scheduled_query_util.delete_s3_bucket(self.ERROR_BUCKET_NAME)

        write_util.delete_table(self.sq_database_name, self.sq_table_name)
        write_util.delete_database(self.sq_database_name)

        write_util.delete_table(self.database_name, self.table_name)
        write_util.delete_database(self.database_name)

File: sample_apps_reinvent2021/javaV2/src/main/java/com/amazonaws/services/timestream/ScheduledQueryExample.java
package com.amazonaws.services.timestream;

import java.time.Instant;
import java.time.LocalDate;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import org.apache.commons.lang3.tuple.Pair;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.S3Object;
import software.amazon.awssdk.services.timestreamquery.TimestreamQueryClient;
import software.amazon.awssdk.services.timestreamquery.model.CreateScheduledQueryRequest;
import software.amazon.awssdk.services.timestreamquery.model.CreateScheduledQueryResponse;
import software.amazon.awssdk.services.timestreamquery.model.DeleteScheduledQueryRequest;
import software.amazon.awssdk.services.timestreamquery.model.DescribeScheduledQueryRequest;
import software.amazon.awssdk.services.timestreamquery.model.DescribeScheduledQueryResponse;
import software.amazon.awssdk.services.timestreamquery.model.DimensionMapping;
import software.amazon.awssdk.services.timestreamquery.model.ErrorReportConfiguration;
import software.amazon.awssdk.services.timestreamquery.model.ExecuteScheduledQueryRequest;
import software.amazon.awssdk.services.timestreamquery.model.ExecuteScheduledQueryResponse;
import software.amazon.awssdk.services.timestreamquery.model.ListScheduledQueriesRequest;
import software.amazon.awssdk.services.timestreamquery.model.ListScheduledQueriesResponse;
import software.amazon.awssdk.services.timestreamquery.model.MeasureValueType;
import software.amazon.awssdk.services.timestreamquery.model.MixedMeasureMapping;
import software.amazon.awssdk.services.timestreamquery.model.MultiMeasureAttributeMapping;
import software.amazon.awssdk.services.timestreamquery.model.MultiMeasureMappings;
import software.amazon.awssdk.services.timestreamquery.model.NotificationConfiguration;
import software.amazon.awssdk.services.timestreamquery.model.ResourceNotFoundException;
import software.amazon.awssdk.services.timestreamquery.model.S3Configuration;
import software.amazon.awssdk.services.timestreamquery.model.ScheduleConfiguration;
import software.amazon.awssdk.services.timestreamquery.model.ScheduledQuery;
import software.amazon.awssdk.services.timestreamquery.model.ScheduledQueryState;
import software.amazon.awssdk.services.timestreamquery.model.SnsConfiguration;
import software.amazon.awssdk.services.timestreamquery.model.TargetConfiguration;
import software.amazon.awssdk.services.timestreamquery.model.TimestreamConfiguration;
import software.amazon.awssdk.services.timestreamquery.model.UpdateScheduledQueryRequest;

import com.amazonaws.services.timestream.utils.ErrorReportHelper;
import com.google.gson.JsonObject;

import static com.amazonaws.services.timestream.CrudAndSimpleIngestionExample.HT_TTL_HOURS;
import static com.amazonaws.services.timestream.Main.DATABASE_NAME;
import static com.amazonaws.services.timestream.Main.TABLE_NAME;
import static software.amazon.awssdk.services.timestreamquery.model.MeasureValueType.DOUBLE;

public class ScheduledQueryExample {
    private final TimestreamQueryClient queryClient;
    public static final String SCHEDULED_QUERY_EXAMPLE = "ScheduledQuerySample";

    public ScheduledQueryExample(TimestreamQueryClient queryClient) {
        this.queryClient = queryClient;
    }

    private final static String HOSTNAME = "host-24Gju";
    private final static String SQ_NAME = "daily-sample";
    /**
     * Timestream ScheduledQuery supports both Fixed Rate expressions and Cron Expressions.
     * 1. SUPPORTED FIXED RATE FORMAT
     *      Syntax: rate(value unit)
     *      - value : A positive number.
     *      - unit : The unit of time. Valid values: minute | minutes | hour | hours | day | days
     *      Examples: rate(1minute), rate(5 day), rate(1 hours)
     * 2. CRON EXPRESSION FORMAT
     *     Cron expressions are comprised of 5 required fields and one optional field
     *     separated by white space. More details: https://tinyurl.com/rxe5zkff
     */
    private static final String SCHEDULE_EXPRESSION = "cron(0/1 * * * ? *)";

    // A Query with a column `timestamp` containing a date that is in the future.
    // The Query as such is valid, but when the result of the query is used as a timeColumn for ingesting into a
    // destination Table, it would fail if the time-range is outside table's retention window (MemoryStoreRetentionPeriodInHours).
    public static String INVALID_QUERY = String.format("SELECT cast('%s' as TIMESTAMP) as timestamp, 1.0 as random_measure_value, " +
            "'1' as dim0", LocalDate.now().plusDays(TimeUnit.HOURS.toDays(HT_TTL_HOURS) + 1));

    // Find the average, p90, p95, and p99 CPU utilization for a specific EC2 host over the past 2 hours.
    public static String VALID_QUERY = "SELECT region, az, hostname, BIN(time, 15s) AS binned_timestamp, " +
            "ROUND(AVG(cpu_utilization), 2) AS avg_cpu_utilization, " +
            "ROUND(APPROX_PERCENTILE(cpu_utilization, 0.9), 2) AS p90_cpu_utilization, " +
            "ROUND(APPROX_PERCENTILE(cpu_utilization, 0.95), 2) AS p95_cpu_utilization, " +
            "ROUND(APPROX_PERCENTILE(cpu_utilization, 0.99), 2) AS p99_cpu_utilization " +
            "FROM " +  DATABASE_NAME + "." +  TABLE_NAME + " " +
            "WHERE measure_name = 'metrics' " +
            "AND hostname = '" + HOSTNAME + "' " +
            "AND time > ago(2h) " +
            "GROUP BY region, hostname, az, BIN(time, 15s) " +
            "ORDER BY binned_timestamp ASC " +
            "LIMIT 5";

    private String createScheduledQuery(String topicArn, String roleArn,
            String s3ErrorReportBucketName, String query, TargetConfiguration targetConfiguration) {
        System.out.println("Creating Scheduled Query");

        CreateScheduledQueryRequest createScheduledQueryRequest = CreateScheduledQueryRequest.builder()
                .name(SQ_NAME)
                .queryString(query)
                .scheduleConfiguration(ScheduleConfiguration.builder()
                        .scheduleExpression(SCHEDULE_EXPRESSION)
                        .build())
                .notificationConfiguration(NotificationConfiguration.builder()
                        .snsConfiguration(SnsConfiguration.builder()
                                .topicArn(topicArn)
                                .build())
                        .build())
                .targetConfiguration(targetConfiguration)
                .errorReportConfiguration(ErrorReportConfiguration.builder()
                        .s3Configuration(S3Configuration.builder()
                                .bucketName(s3ErrorReportBucketName)
                                .objectKeyPrefix(SCHEDULED_QUERY_EXAMPLE)
                                .build())
                        .build())
                .scheduledQueryExecutionRoleArn(roleArn)
                .build();

        try {
            final CreateScheduledQueryResponse response = queryClient.createScheduledQuery(createScheduledQueryRequest);
            final String scheduledQueryArn = response.arn();
            System.out.println("Successfully created scheduled query : " + scheduledQueryArn);
            return scheduledQueryArn;
        }
        catch (Exception e) {
            System.out.println("Scheduled Query creation failed: " + e);
            throw e;
        }
    }

    public String createValidScheduledQuery(String topicArn, String roleArn,
            String databaseName, String tableName, String s3ErrorReportBucketName) {
        List<Pair<String, MeasureValueType>> sourceColToMeasureValueTypes = Arrays.asList(
                Pair.of("avg_cpu_utilization", DOUBLE),
                Pair.of("p90_cpu_utilization", DOUBLE),
                Pair.of("p95_cpu_utilization", DOUBLE),
                Pair.of("p99_cpu_utilization", DOUBLE));

        TargetConfiguration targetConfiguration = TargetConfiguration.builder()
                .timestreamConfiguration(TimestreamConfiguration.builder()
                .databaseName(databaseName)
                .tableName(tableName)
                .timeColumn("binned_timestamp")
                .dimensionMappings(Arrays.asList(
                        DimensionMapping.builder()
                                .name("region")
                                .dimensionValueType("VARCHAR")
                                .build(),
                        DimensionMapping.builder()
                                .name("az")
                                .dimensionValueType("VARCHAR")
                                .build(),
                        DimensionMapping.builder()
                                .name("hostname")
                                .dimensionValueType("VARCHAR")
                                .build()
                ))
                .multiMeasureMappings(MultiMeasureMappings.builder()
                        .targetMultiMeasureName("multi-metrics")
                        .multiMeasureAttributeMappings(
                                sourceColToMeasureValueTypes.stream()
                                        .map(pair -> MultiMeasureAttributeMapping.builder()
                                                .measureValueType(pair.getValue().name())
                                                .sourceColumn(pair.getKey())
                                                .build())
                                        .collect(Collectors.toList()))
                        .build())
                .build())
                .build();

        return createScheduledQuery(topicArn, roleArn, s3ErrorReportBucketName, VALID_QUERY, targetConfiguration);
    }

    public String createInvalidScheduledQuery(String topicArn, String roleArn,
            String databaseName, String tableName, String s3ErrorReportBucketName) {

        TargetConfiguration targetConfiguration = TargetConfiguration.builder()
                .timestreamConfiguration(TimestreamConfiguration.builder()
                        .databaseName(databaseName)
                        .tableName(tableName)
                        .timeColumn("timestamp")
                        .dimensionMappings(Collections.singletonList(
                                DimensionMapping.builder()
                                        .name("dim0")
                                        .dimensionValueType("VARCHAR")
                                        .build()
                        ))
                        .mixedMeasureMappings(MixedMeasureMapping.builder()
                                .sourceColumn("random_measure_value")
                                .measureValueType(DOUBLE)
                                .build())
                        .build())
                .build();

        return createScheduledQuery(topicArn, roleArn, s3ErrorReportBucketName, INVALID_QUERY, targetConfiguration);
    }

    public void deleteScheduledQuery(String scheduledQueryArn) {
        System.out.println("Deleting Scheduled Query");

        try {
            queryClient.deleteScheduledQuery(DeleteScheduledQueryRequest.builder()
                    .scheduledQueryArn(scheduledQueryArn).build());
            System.out.println("Successfully deleted scheduled query");
        }
        catch (Exception e) {
            System.out.println("Scheduled Query deletion failed: " + e);
            //do not throw exception here, because we want the following cleanups executes
        }
    }

    public void listScheduledQueries() {
        System.out.println("Listing Scheduled Queries");
        try {
            String nextToken = null;

            do {
                ListScheduledQueriesResponse listScheduledQueriesResult =
                        queryClient.listScheduledQueries(ListScheduledQueriesRequest.builder()
                                .nextToken(nextToken).maxResults(10)
                                .build());
                List<ScheduledQuery> scheduledQueryList = listScheduledQueriesResult.scheduledQueries();

                printScheduledQuery(scheduledQueryList);
                nextToken = listScheduledQueriesResult.nextToken();
            } while (nextToken != null);
        }
        catch (Exception e) {
            System.out.println("List Scheduled Queries failed: " + e);
            throw e;
        }
    }

    public void describeScheduledQueries(String scheduledQueryArn) {
        System.out.println("Describing Scheduled Query");
        try {
            DescribeScheduledQueryResponse describeScheduledQueryResult =
                    queryClient.describeScheduledQuery(DescribeScheduledQueryRequest.builder()
                            .scheduledQueryArn(scheduledQueryArn)
                            .build());
            System.out.println(describeScheduledQueryResult);
        }
        catch (ResourceNotFoundException e) {
            System.out.println("Scheduled Query doesn't exist");
            throw e;
        }
        catch (Exception e) {
            System.out.println("Describe Scheduled Query failed: " + e);
            throw e;
        }
    }

    public void executeScheduledQuery(String scheduledQueryArn) {
        System.out.println("Running Scheduled Query");
        try {
            ExecuteScheduledQueryResponse executeScheduledQueryResult = queryClient.executeScheduledQuery(ExecuteScheduledQueryRequest.builder()
                    .scheduledQueryArn(scheduledQueryArn)
                    .invocationTime(Instant.now())
                    .build()
            );

            System.out.println("ScheduledQuery run response code: " + executeScheduledQueryResult.sdkHttpResponse().statusCode());

        }
        catch (ResourceNotFoundException e) {
            System.out.println("Scheduled Query doesn't exist");
            throw e;
        }
        catch (Exception e) {
            System.out.println("Scheduled Query run failed: " + e);
            throw e;
        }
    }

    public void updateScheduledQuery(String scheduledQueryArn, ScheduledQueryState state) {
        System.out.println("Updating Scheduled Query");
        try {
            queryClient.updateScheduledQuery(UpdateScheduledQueryRequest.builder()
                    .scheduledQueryArn(scheduledQueryArn)
                    .state(state)
                    .build());
            System.out.println("Successfully updated scheduled query state");
        }
        catch (ResourceNotFoundException e) {
            System.out.println("Scheduled Query doesn't exist");
            throw e;
        }
        catch (Exception e) {
            System.out.println("Scheduled Query update failed: " + e);
            throw e;
        }
    }

    public void printScheduledQuery(List<ScheduledQuery> scheduledQueryList) {
        for (ScheduledQuery scheduledQuery: scheduledQueryList) {
            System.out.println(scheduledQuery.arn());
        }
    }

    public void parseS3ErrorReport(final S3Client s3Client, final String s3ErrorReportBucketName,
            final String errorReportPrefix) {
        List<S3Object> s3Objects = ErrorReportHelper.listReportFilesInS3(s3Client, s3ErrorReportBucketName, errorReportPrefix);
        JsonObject errorReport = ErrorReportHelper.readErrorReport(s3Client, s3Objects.get(0), s3ErrorReportBucketName);
        System.out.println("Error report from S3:: \n" + errorReport);
        // Since we expect only one error, print the first one
        System.out.println("Error reason from S3:: " +
                errorReport.get("errors").getAsJsonArray().get(0)
                        .getAsJsonObject().get("reason").getAsString());
    }

    public enum NotificationType {
        SCHEDULED_QUERY_CREATING("SCHEDULED_QUERY_CREATING"), // scheduled query is being created
        SCHEDULED_QUERY_CREATED("SCHEDULED_QUERY_CREATED"), // for successful scheduled query creation
        SCHEDULED_QUERY_UPDATED("SCHEDULED_QUERY_UPDATED"), // for successful scheduled query update
        SCHEDULED_QUERY_DELETED("SCHEDULED_QUERY_DELETED"), // for successful scheduled query deletion
        AUTO_TRIGGER_SUCCESS("AUTO_TRIGGER_SUCCESS"), // for successful completion of scheduled query
        AUTO_TRIGGER_FAILURE("AUTO_TRIGGER_FAILURE"), // for failures of scheduled query
        MANUAL_TRIGGER_SUCCESS("MANUAL_TRIGGER_SUCCESS"), // for successful completion of manual execution
        MANUAL_TRIGGER_FAILURE("MANUAL_TRIGGER_FAILURE"); // for failed manual execution

        String value;

        NotificationType(String value) {
            this.value = value;
        }
    }
}
